File: backend/src/modules/users/controllers/systemUsersController.ts
import { Request, Response, NextFunction } from 'express'
import { prisma } from '../../../config/database'
import { AppError } from '../../../shared/errors/AppError'
import * as bcrypt from 'bcrypt'

const DEFAULT_PERMISSIONS: any = {
  ADMIN: {
    'agenda.view': true,
    'agenda.manage': true,
    'patients.view': true,
    'patients.manage': true,
    'billing.view': true,
    'billing.manage': true,
    'settings.view': true,
    'settings.manage': true,
    'chat.access': true,
  },
  STAFF: {
    'agenda.view': true,
    'agenda.manage': true,
    'patients.view': true,
    'patients.manage': true,
    'billing.view': false,
    'billing.manage': false,
    'settings.view': false,
    'settings.manage': false,
    'chat.access': true,
  },
  MEMBER: {
    'agenda.view': true,
    'agenda.manage': false,
    'patients.view': true,
    'patients.manage': false,
    'billing.view': false,
    'billing.manage': false,
    'settings.view': false,
    'settings.manage': false,
    'chat.access': true,
  }
}

export class SystemUsersController {
  async index(req: Request, res: Response, next: NextFunction) {
    try {
      const { clinicId } = req.query as { clinicId?: string }

      const systemUsers = await prisma.systemUser.findMany({
        where: clinicId ? { clinicId } : {},
        include: {
          user: {
            select: { id: true, name: true, email: true, active: true, phone: true }
          }
        },
        orderBy: { createdAt: 'desc' }
      })
      res.json(systemUsers)
    } catch (err) {
      next(err)
    }
  }

  async create(req: Request, res: Response, next: NextFunction) {
    try {
      const { clinicId, name, email, role, password, permissions } = req.body

      // Try to find an existing user by email, falling back to creating a new one.
      // Using findFirst since email is no longer unique.
      let user = await prisma.user.findFirst({ where: { email } })
      if (!user) {
        const rawPassword = password?.trim() || '123456'
        const passwordHash = await bcrypt.hash(rawPassword, 8)
        user = await prisma.user.create({
          data: {
            name,
            email,
            passwordHash,
            role: 'STAFF',
          }
        })
      }

      const systemUser = await prisma.systemUser.create({
        data: {
          clinicId,
          userId: user.id,
          role,
          active: true,
          permissions: permissions || DEFAULT_PERMISSIONS[role] || DEFAULT_PERMISSIONS.STAFF
        },
        include: { user: { select: { id: true, name: true, email: true, active: true } } }
      })

      res.status(201).json(systemUser)
    } catch (err) {
      next(err)
    }
  }

  async update(req: Request, res: Response, next: NextFunction) {
    try {
      const { id } = req.params as { id: string }
      const { role, active, permissions } = req.body

      const updated = await prisma.systemUser.update({
        where: { id },
        data: { role, active, permissions },
        include: { user: { select: { id: true, name: true, email: true, active: true } } }
      })

      res.json(updated)
    } catch (err) {
      next(err)
    }
  }

  async remove(req: Request, res: Response, next: NextFunction) {
    try {
      const { id } = req.params as { id: string }

      const systemUser = await prisma.systemUser.findUnique({ where: { id } })
      if (!systemUser) {
        throw new AppError('Usuário não encontrado na clínica', 404)
      }

      await prisma.systemUser.delete({ where: { id } })

      res.json({ message: 'Usuário removido da clínica com sucesso' })
    } catch (err) {
      next(err)
    }
  }
}

File: backend/Consultorio.API/Controllers/SystemUsersController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Consultorio.Domain.Models;
using Consultorio.Infra.Context;

namespace Consultorio.API.Controllers;

[ApiController]
[Route("api/system-users")]
[Authorize]
public class SystemUsersController : ControllerBase
{
    private readonly AppDbContext _db;

    public SystemUsersController(AppDbContext db) => _db = db;

    private Guid GetClinicId() =>
        Guid.TryParse(User.FindFirst("clinicId")?.Value, out var id) ? id : Guid.Empty;

    [HttpGet]
    public async Task<ActionResult> GetAll()
    {
        var clinicId = GetClinicId();
        if (clinicId == Guid.Empty)
            return BadRequest(new { message = "Clínica não identificada." });

        var users = await _db.SystemUsers
            .Include(su => su.User)
            .Where(su => su.ClinicId == clinicId)
            .OrderBy(su => su.User.Name)
            .ToListAsync();

        return Ok(users.Select(ToResponse));
    }

    [HttpPost]
    public async Task<ActionResult> Create([FromBody] CreateSystemUserDto dto)
    {
        var clinicId = dto.ClinicId != Guid.Empty ? dto.ClinicId : GetClinicId();
        if (clinicId == Guid.Empty)
            return BadRequest(new { message = "Clínica não identificada." });

        if (string.IsNullOrWhiteSpace(dto.Name) || string.IsNullOrWhiteSpace(dto.Email))
            return BadRequest(new { message = "Nome e e-mail são obrigatórios." });

        var email = dto.Email.Trim().ToLowerInvariant();

        var password = string.IsNullOrWhiteSpace(dto.Password)
            ? "123456"
            : dto.Password.Trim();

        var role = NormalizeRole(dto.Role);

        var user = new User
        {
            Id = Guid.NewGuid(),
            Name = dto.Name.Trim(),
            Email = email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(password),
            Phone = dto.Phone?.Trim(),
            IsActive = dto.Active,
            CreatedAt = DateTime.UtcNow,
        };

        var systemUser = new SystemUser
        {
            Id = Guid.NewGuid(),
            UserId = user.Id,
            ClinicId = clinicId,
            Role = role,
            CreatedAt = DateTime.UtcNow,
        };

        _db.Users.Add(user);
        _db.SystemUsers.Add(systemUser);
        await _db.SaveChangesAsync();

        var response = ToResponse(systemUser, user);
        response.GeneratedPassword = password;
        return Ok(response);
    }

    [HttpPut("{id}")]
    public async Task<ActionResult> Update(Guid id, [FromBody] UpdateSystemUserDto dto)
    {
        var systemUser = await _db.SystemUsers
            .Include(su => su.User)
            .FirstOrDefaultAsync(su => su.Id == id);

        if (systemUser == null)
            return NotFound(new { message = "Usuário não encontrado." });

        if (!string.IsNullOrWhiteSpace(dto.Name))
            systemUser.User.Name = dto.Name.Trim();

        if (!string.IsNullOrWhiteSpace(dto.Email))
        {
            systemUser.User.Email = dto.Email.Trim().ToLowerInvariant();
        }

        if (!string.IsNullOrWhiteSpace(dto.Role))
            systemUser.Role = NormalizeRole(dto.Role);

        if (dto.Active.HasValue)
            systemUser.User.IsActive = dto.Active.Value;

        if (!string.IsNullOrWhiteSpace(dto.Phone))
            systemUser.User.Phone = dto.Phone.Trim();

        // Troca de senha por admin: hash com BCrypt antes de salvar.
        // Mínimo de 6 caracteres é validado aqui mesmo para evitar
        // senhas excessivamente fracas vindas da UI.
        if (!string.IsNullOrWhiteSpace(dto.Password))
        {
            var pwd = dto.Password.Trim();
            if (pwd.Length < 6)
                return BadRequest(new { message = "A senha deve ter ao menos 6 caracteres." });
            systemUser.User.PasswordHash = BCrypt.Net.BCrypt.HashPassword(pwd);
        }

        systemUser.UpdatedAt = DateTime.UtcNow;
        systemUser.User.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        return Ok(ToResponse(systemUser));
    }

    [HttpDelete("{id}")]
    public async Task<ActionResult> Delete(Guid id)
    {
        var systemUser = await _db.SystemUsers.FindAsync(id);

        if (systemUser == null)
            return NotFound(new { message = "Usuário não encontrado." });

        _db.SystemUsers.Remove(systemUser);
        await _db.SaveChangesAsync();

        return NoContent();
    }

    private static SystemUserResponseDto ToResponse(SystemUser su) => ToResponse(su, su.User);

    private static SystemUserResponseDto ToResponse(SystemUser su, User user) => new()
    {
        Id = su.Id,
        ClinicId = su.ClinicId,
        UserId = su.UserId,
        Role = su.Role,
        Active = user.IsActive,
        CreatedAt = su.CreatedAt,
        User = new SystemUserUserDto
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            Phone = user.Phone,
            Active = user.IsActive,
        },
        Permissions = new Dictionary<string, bool>(),
    };

    private static string NormalizeRole(string? role)
    {
        return string.Equals(role, "ADMIN", StringComparison.OrdinalIgnoreCase)
            ? "ADMIN"
            : "RECEPTIONIST";
    }

}

public class CreateSystemUserDto
{
    public Guid ClinicId { get; set; }
    public string Name { get; set; } = null!;
    public string Email { get; set; } = null!;
    public string Role { get; set; } = "RECEPTIONIST";
    public bool Active { get; set; } = true;
    public string? Password { get; set; }
    public string? Phone { get; set; }
    public Dictionary<string, bool>? Permissions { get; set; }
}

public class UpdateSystemUserDto
{
    public string? Name { get; set; }
    public string? Email { get; set; }
    public string? Phone { get; set; }
    public string? Role { get; set; }
    public bool? Active { get; set; }
    public string? Password { get; set; }
    public Dictionary<string, bool>? Permissions { get; set; }
}

public class SystemUserResponseDto
{
    public Guid Id { get; set; }
    public Guid ClinicId { get; set; }
    public Guid UserId { get; set; }
    public string Role { get; set; } = null!;
    public bool Active { get; set; }
    public DateTime CreatedAt { get; set; }
    public string? GeneratedPassword { get; set; }
    public SystemUserUserDto User { get; set; } = null!;
    public Dictionary<string, bool> Permissions { get; set; } = new();
}

public class SystemUserUserDto
{
    public Guid Id { get; set; }
    public string Name { get; set; } = null!;
    public string Email { get; set; } = null!;
    public string? Phone { get; set; }
    public bool Active { get; set; }
}
